File: solutions/csharp/complex-numbers/1/ComplexNumbers.cs
using System;

public struct ComplexNumber
{
    private double r;
    private double i;
    public ComplexNumber(double real, double imaginary)
    {
        r = real;
        i = imaginary;
    }

    public double Real() => r;

    public double Imaginary() => i;

    public ComplexNumber Mul(ComplexNumber other)
    {
        ComplexNumber t = new ComplexNumber();
        t.r = r * other.r - i * other.i;
        t.i = i * other.r + r * other.i;
        return t;
    }


    public ComplexNumber Add(ComplexNumber other)
    {
        ComplexNumber t = new ComplexNumber();
        t.r = r + other.r;
        t.i = i + other.i;
        return t;
    }

    public ComplexNumber Sub(ComplexNumber other)
    {
        ComplexNumber t = new ComplexNumber();
        t.r = r - other.r;
        t.i = i - other.i;
        return t;
    }

    public ComplexNumber Div(ComplexNumber other)
    {
        ComplexNumber t = new ComplexNumber();
        t.r = (r * other.r + i * other.i)/(other.r * other.r + other.i * other.i);
        t.i = (i * other.r - r * other.i)/(other.r * other.r + other.i * other.i);
        return t;
    }

    public double Abs()
    {
        return Math.Sqrt(r*r + i *i);
    }

    public ComplexNumber Conjugate()
    {
        ComplexNumber t = new ComplexNumber();
        t.r = r;
        t.i = -1 * i;
        return t;
    }

    public ComplexNumber Exp()
    {
        ComplexNumber t = new ComplexNumber();
        t.r = Math.Exp(r) * Math.Cos(i);
        t.i = Math.Exp(r) * Math.Sin(i);
        return t;
    }
}

File: solutions/python/complex-numbers/1/complex_numbers.py
import math
class ComplexNumber:
    def __init__(self, real, imaginary):
        self.real=real
        self.imaginary=imaginary
       
    def __eq__(self, other):
        if type(other) is not ComplexNumber:
            return self.real == other
        else:
            return self.real == other.real and self.imaginary == other.imaginary

    def __add__(self, other):
        self.real += other.real
        self.imaginary += other.imaginary
        return(self)
        

    def __mul__(self, other):
        cmpl=ComplexNumber(0,0)
        cmpl.real = self.real * other.real - self.imaginary * other.imaginary 
        cmpl.imaginary = self.imaginary * other.real + self.real * other.imaginary
        return(cmpl)

    def __sub__(self, other):
        self.real -= other.real
        self.imaginary -= other.imaginary
        return(self)

    def __truediv__(self, other
    ):
        cmpl=ComplexNumber(0,0)
        cmpl.real = (self.real * other.real + self.imaginary * other.imaginary) / \
                    (other.real * other.real + other.imaginary * other.imaginary) 
        cmpl.imaginary = (self.imaginary * other.real - self.real * other.imaginary) / \
                    (other.real * other.real + other.imaginary * other.imaginary)
        return(cmpl)

    def __abs__(self):
        self.real = math.sqrt(self.real * self.real + self.imaginary * self.imaginary)
        del self.imaginary
        return(self)

    def conjugate(self):
        self.imaginary = -self.imaginary
        return(self)

    def exp(self):
        self = ComplexNumber(math.exp(self.real), 0) * \
               ComplexNumber(math.cos(self.imaginary),math.sin(self.imaginary))       
        return(self)

    def __round__(self,ndigs):
        self.real=round(self.real,ndigs)
        if hasattr(self,'imaginary'):
            self.imaginary=round(self.imaginary,ndigs)
        return(self)
